File: smtk/attribute/PythonTesting/expressionTest2.py
"""
  Manual port of SMTK/smtk/attribute/Testing/expressionTest.cxx
  For verifying python-shiboken wrappers
  
  Requires SMTKCorePython.so to be in module path
  """

import smtk

def addItemDefinition( ato, data_type, name):
  def_ = data_type.New(name)
  if def_ is None:
    print "could not create"
    return None
  idef = data_type.ToItemDefinition(def_)
  if idef is None:
    print "could not convert"
    return None
  if not ato.addItemDefinition(idef):
    print "could not add"
    return None
  return def_


if __name__ == '__main__':
  import sys
  
  status = 0
  
  manager = smtk.attribute.Manager()
  print 'Manager created'
  
  # Lets create an attribute to represent an expression
  expDef = manager.createDefinition("ExpDef");
  eitemdef = addItemDefinition(expDef, smtk.attribute.StringItemDefinition, "Expression String");
  eitemdef2 = addItemDefinition( expDef, smtk.attribute.StringItemDefinition, "Aux String");
  eitemdef.setDefaultValue("sample");
  
  base = manager.createDefinition("BaseDef");
  #Lets add some item definitions
  iitemdef = addItemDefinition( base, smtk.attribute.IntItemDefinition, "IntItem1");
  iitemdef = addItemDefinition( base, smtk.attribute.IntItemDefinition, "IntItem2");
  iitemdef.setDefaultValue(10);

  def1 = manager.createDefinition("Derived1", "BaseDef");
  # add some item definitions
  ditemdef = addItemDefinition(def1, smtk.attribute.DoubleItemDefinition, "DoubleItem1");
  # Allow this one to hold an expression
  ditemdef.setExpressionDefinition(expDef);
  # Check to make sure we can use expressions
  if not ditemdef.allowsExpressions():
    print  "ERROR - Item Def does not allow expressions";
    status = -1;
  ditemdef = addItemDefinition(def1, smtk.attribute.DoubleItemDefinition, "DoubleItem2");
  ditemdef.setDefaultValue(-35.2);

  def2 = manager.createDefinition("Derived2", "Derived1");
  # Lets add some item definitions
  sitemdef = addItemDefinition(def2, smtk.attribute.StringItemDefinition, "StringItem1");
  sitemdef = addItemDefinition(def2, smtk.attribute.StringItemDefinition, "StringItem2");
  sitemdef.setDefaultValue("Default");
  
  # Lets test creating an attribute by passing in the expression definition explicitly
  expAtt = manager.createAttribute("Exp1", expDef);
  att = manager.createAttribute("testAtt", "Derived2");
  if att is not None:
      print "Attribute testAtt created";
  else:
      print "ERROR: Attribute testAtt not created";
      status = -1;

  # Find the expression enabled item
  vitem = smtk.attribute.ValueItem.CastTo(att.item(2))
  if vitem.allowsExpressions():
    vitem.setExpression(expAtt);
    print "Expression Set on %s" % vitem.name()
  else:
    print "ERROR: Can not set expression on %s" % vitem.name();
    status = -1

  n = att.numberOfItems();
  print "Items of testAtt:";
  for i in range(0,n,1):
    item = att.item(i);
    sys.stdout.write("\t%s Type = %s, " % (item.name(), smtk.attribute.Item.type2String(item.type())))
    vitem = smtk.attribute.ValueItem.CastTo(item);
    if vitem is not None:
      if vitem.isExpression(0):
        print " using Expression: %s" % vitem.expression(0).name();
      else:
        print " Value = %s" % vitem.valueAsString();

  print
  input = raw_input('Hit <Enter> to delete Managar and exit')
  del manager
  print 'Manager destroyed'
  
  sys.exit(status)

File: smtk/attribute/testing/cxx/expressionTest2.cxx
//=========================================================================
//  Copyright (c) Kitware, Inc.
//  All rights reserved.
//  See LICENSE.txt for details.
//
//  This software is distributed WITHOUT ANY WARRANTY; without even
//  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
//  PURPOSE.  See the above copyright notice for more information.
//=========================================================================

#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/Definition.h"
#include "smtk/attribute/DoubleItem.h"
#include "smtk/attribute/DoubleItemDefinition.h"
#include "smtk/attribute/IntItem.h"
#include "smtk/attribute/IntItemDefinition.h"
#include "smtk/attribute/StringItem.h"
#include "smtk/attribute/StringItemDefinition.h"
#include "smtk/attribute/System.h"
#include <iostream>

int main()
{
  int status = 0;
  {
    smtk::attribute::System system;
    std::cout << "System Created\n";
    // Lets create an attribute to represent an expression
    smtk::attribute::DefinitionPtr expDef = system.createDefinition("ExpDef");
    smtk::attribute::StringItemDefinitionPtr eitemdef =
      expDef->addItemDefinition<smtk::attribute::StringItemDefinitionPtr>("Expression String");
    smtk::attribute::StringItemDefinitionPtr eitemdef2 =
      expDef->addItemDefinition<smtk::attribute::StringItemDefinition>("Aux String");
    eitemdef->setDefaultValue("sample");

    smtk::attribute::DefinitionPtr base = system.createDefinition("BaseDef");
    // Lets add some item definitions
    smtk::attribute::IntItemDefinitionPtr iitemdef =
      base->addItemDefinition<smtk::attribute::IntItemDefinitionPtr>("IntItem1");
    iitemdef = base->addItemDefinition<smtk::attribute::IntItemDefinitionPtr>("IntItem2");
    iitemdef->setDefaultValue(10);

    smtk::attribute::DefinitionPtr def1 = system.createDefinition("Derived1", "BaseDef");
    // Lets add some item definitions
    smtk::attribute::DoubleItemDefinitionPtr ditemdef =
      def1->addItemDefinition<smtk::attribute::DoubleItemDefinitionPtr>("DoubleItem1");
    // Allow this one to hold an expression
    ditemdef->setExpressionDefinition(expDef);
    // Check to make sure we can use expressions
    if (!ditemdef->allowsExpressions())
    {
      std::cout << "ERROR - Item Def does not allow expressions\n";
      status = -1;
    }
    ditemdef = def1->addItemDefinition<smtk::attribute::DoubleItemDefinitionPtr>("DoubleItem2");
    ditemdef->setDefaultValue(-35.2);

    smtk::attribute::DefinitionPtr def2 = system.createDefinition("Derived2", "Derived1");
    // Lets add some item definitions
    smtk::attribute::StringItemDefinitionPtr sitemdef =
      def2->addItemDefinition<smtk::attribute::StringItemDefinitionPtr>("StringItem1");
    sitemdef = def2->addItemDefinition<smtk::attribute::StringItemDefinitionPtr>("StringItem2");
    sitemdef->setDefaultValue("Default");

    // Lets test creating an attribute by passing in the expression definition explicitly
    smtk::attribute::AttributePtr expAtt = system.createAttribute("Exp1", expDef);
    smtk::attribute::AttributePtr att = system.createAttribute("testAtt", "Derived2");
    if (att)
    {
      std::cout << "Attribute testAtt created\n";
    }
    else
    {
      std::cout << "ERROR: Attribute testAtt not created\n";
      status = -1;
    }

    smtk::attribute::ValueItemPtr vitem;
    smtk::attribute::ItemPtr item;

    // Find the expression enabled item
    item = att->item(2);
    vitem = smtk::dynamic_pointer_cast<smtk::attribute::ValueItem>(item);
    if (vitem->allowsExpressions())
    {
      vitem->setExpression(expAtt);
      std::cout << "Expression Set on " << vitem->name() << "\n";
    }
    else
    {
      std::cout << "ERROR: Can not set expression on " << vitem->name() << "\n";
      status = -1;
    }

    int i, n = static_cast<int>(att->numberOfItems());
    std::cout << "Items of testAtt:\n";
    for (i = 0; i < n; i++)
    {
      item = att->item(i);
      std::cout << "\t" << item->name()
                << " Type = " << smtk::attribute::Item::type2String(item->type()) << ", ";
      vitem = smtk::dynamic_pointer_cast<smtk::attribute::ValueItem>(item);
      if (vitem)
      {
        if (vitem->isExpression())
        {
          std::cout << " using Expression: " << vitem->expression()->name() << "\n";
        }
        else
        {
          std::cout << " Value = " << vitem->valueAsString() << "\n";
        }
      }
    }
    std::cout << "System destroyed\n";
  }
  return status;
}
